crypto/evp/kdf_meth.c

```diff
@@ -80,6 +80,8 @@ static void *evp_kdf_from_algorithm(int name_id,
             if (kdf->newctx != NULL)
                 break;
             kdf->newctx = OSSL_FUNC_kdf_newctx(fns);
+            if (kdf->newctx == NULL)
+                break;
```

Comment:
Do we really want to add these checks? If so, should we add them to all _meth.c files that fetch functions from dispatch arrays?? This is to guard against provider authors errors but I see it highly unlikely that someone would bother adding the entry in the dispatch table but with NULL function pointer.

Reply:
I agree, we probably should add them to all meth.c files, I only limited this to the kdf method as thats what the issue described.  I assert we should add them because the docs currently mandate that the context handling functions are required, but we don't check for them.  Only looked at the kdf docs mind you, not sure what the other methods mandate

Reply:
I'd split this into a separate PR and add the NULL checks in all those from_dispatch() calls then. IMO it is inconsistent to add such NULL checks only to the kdf from_dispatch call.

Reply:
So i went through and audited the dispatch function loading for the various provider types we have.  As it happens, we only have three providers that don't require any specific methods (the encoder, decoder and storemgmt providers).  The remaining providers all make checks for the minimum required dispatch functions (which unfortunately aren't documented).  

I suppose the path forward here then, is to, in a separate PR, document the dispatch methods that are required for each provider, and in the case of the three with no checks, add whatever we think the minimum requirement is.

@t8m do you have any insight as to what the required minimum for those providers are?

Reply:
Hmm, I do not know which functions for these are mandatory. @levitte ?

Reply:
> ... we only have three providers that don't require any specific methods (the encoder, decoder and storemgmt providers)

Er, this isn't quite right.  They do require specific methods, they just don't use the counting method.

https://github.com/openssl/openssl/blob/5ed9a32a2aee89e10eb2891f5fb7a283e1b5199b/crypto/store/store_meth.c#L233-L237
https://github.com/openssl/openssl/blob/5ed9a32a2aee89e10eb2891f5fb7a283e1b5199b/crypto/encode_decode/decoder_meth.c#L265-L276
https://github.com/openssl/openssl/blob/5ed9a32a2aee89e10eb2891f5fb7a283e1b5199b/crypto/encode_decode/encoder_meth.c#L273-L286

The common criteria for picking what minimum set of functions is mandatory is some sort of "common sense", and it's really about what it would make sense for a provider author to get away with.
In libcrypto, there are numerous places where we check if a specific function pointer isn't NULL before use, and either fall back to some bread-and-butter code or raise some sort of error like "Unsupported".  That's an indication of what functions are optional vs mandatory.

The underlying reasoning is to avoid adding on unnecessary churn on the provider authors if we can help it.  Seen from the other end, making a specific function mandatory would require some amount of reasonable justification (we don't have a very high bar for that, quite frankly, other than at least have put some thought into it).

Reply:
However, we really have two discussions going on here:

1. what functions are considered mandatory and should be checked with that in mind?
2. what to do when an OSSL_DISPATCH entry has a (faulty) NULL function pointer?

The second question is actually something that I don't think we actually considered.  Like @t8m said, who would make such an OSSL_DISPATCH entry?  That being said, bugs happen, and it might be a bit harsh if the program that tries to load a provider *crashes* because the loaded provider has this sort of bug.  So in my mind, checking if every OSSL_DISPATCH entry actually has a non-NULL function pointer is a sane thing to do.
I would do that *before* the switch statement, 'cause it's a bit silly to do that exact same check in every case.